### PR TITLE
nodePackages: Switch to 12.x and 14.x, continue #85764

### PR DESCRIPTION
**Pull request by calbrecht**
###### Motivation for this change
To continue with PR #85764 and implement suggested changes on top of that.

To be consistent with future updates of `Node.js`, i propose (and implement within this PR) on top of the suggested changes in mentionend PR, that we stick to the official released `Node.js` versions `LTS` and `Current`, *now* at 12.x and 14.x.

Let us reflect that in the change of the `nodePackages` attribute in the `top-level/all-packages.nix` package set by renaming the version suffixed package set to `nodePackages_latest` for the latest official `Current` release, and `nodePackages` for the latest official `LTS` release. 

###### Things done

To ensure all node-packages build successfully, i wrote a `shell.nix` for testing.

<details><summary>shell.nix in nixpkgs before PR (hacky, but you get the point)</summary>

```nix
{ pkgs ? import ./default.nix {} }:

let
  node10Pkgs = (import ./pkgs/development/node-packages/default-v10.nix {
    inherit pkgs;
    inherit (pkgs) stdenv;
    nodejs = pkgs.nodejs;
  });
  node12Pkgs = (import ./pkgs/development/node-packages/default-v12.nix {
    inherit pkgs;
    inherit (pkgs) stdenv;
    nodejs = pkgs.nodejs-12_x;
  });
  node13Pkgs = (import ./pkgs/development/node-packages/default-v13.nix {
    inherit pkgs;
    inherit (pkgs) stdenv;
    nodejs = pkgs.nodejs-13_x;
  });
in
pkgs.mkShell {

  buildInputs = with pkgs.lib;
  (attrValues (
    filterAttrs (name: _: 1 > length (
      intersectLists [ name ] [
        # broken packages 10_x
        "coc-imselect"
        "fast-cli-1.x"
        "git-ssb"
        "pulp"
        "insect"
        "node-red"
        "node-inspector"
        "stf"
        "ssb-server"
        # not applicable 10_x
        "ios-deploy"
      ])) node10Pkgs
  )) ++
  (attrValues (
    filterAttrs (name: _: 1 > length (
      intersectLists [ name ] [
        # broken packages 12_x
        # not applicable 12_x
      ])) node12Pkgs
  )) ++
  (attrValues (
    filterAttrs (name: _: 1 > length (
      intersectLists [ name ] [
        # broken packages 13_x
        # not applicable 13_x
      ])) node13Pkgs
  ));

  shellHook = ''
    ${pkgs.lib.concatMapStrings (name: ''
      echo -n 10: ${name}:\ ; echo -en $(which ${name} 2>/dev/null)"\n" ;
    '') (pkgs.lib.attrNames node10Pkgs) }
    ${pkgs.lib.concatMapStrings (name: ''
      echo -n 12: ${name}:\ ; echo -en $(which ${name} 2>/dev/null)"\n" ;
    '') (pkgs.lib.attrNames node12Pkgs) }
    ${pkgs.lib.concatMapStrings (name: ''
      echo -n 13: ${name}:\ ; echo -en $(which ${name} 2>/dev/null)"\n" ;
    '') (pkgs.lib.attrNames node13Pkgs) }
  '';
}
```
</details>

\
With the help of this `shell.nix` i noticed there are some packages already broken in the old `node-packages`. Namely `coc-imselect` `fast-cli-1.x` `git-ssb` `pulp` `insect` `node-red` `node-inspector` `stf` `ssb-server` and one package that claims to be incompatible with `linux`, namely `ios-deploy`.

\
So i thought about what to do if node-packages break with new releases and in different versions, and i came to the conclusion the easiest way is to mark them as `broken` in the `meta` attribute depending on the `nodejs.version` in use. Similar method is applicable to platform support with the `meta.platforms` attribute. So we are able to write in `node-packages/default.nix` e.g.

```nix
let
  since = (version: pkgs.lib.versionAtLeast nodejs.version version);
in nodePackages // {
  # ....
  "fast-cli-1.x" = nodePackages."fast-cli-1.x".override {
    meta.broken = since "10";
  };
}
```
\
To ensure that is working i came up with this `shell.nix`

<details><summary>shell.nix in nixpkgs after PR</summary>

```nix
{ pkgs ? import ./default.nix {} }:

let
  isBroken = pkg: pkgs.lib.hasAttrByPath ["meta" "broken"] pkg && pkg.meta.broken;
  isAvailable = pkg: pkgs.lib.hasAttrByPath ["meta" "available"] pkg && pkg.meta.available;
  toInputs = nodePkgs: pkgs.lib.mapAttrsToList (_: pkg: pkg) nodePkgs;
  filterPkgs = nodePkgs: pkgs.lib.filterAttrs (_: pkg: ! isBroken pkg && isAvailable pkg) nodePkgs;
  toFilteredInputs = nodePkgs: toInputs (filterPkgs nodePkgs);

  node12Pkgs = pkgs.lib.filterAttrs (_: v: pkgs.lib.isDerivation v) pkgs.nodePackages;
  node14Pkgs = pkgs.lib.filterAttrs (_: v: pkgs.lib.isDerivation v) pkgs.nodePackages_latest;
in
pkgs.mkShell {

  buildInputs = (toFilteredInputs node12Pkgs) ++ (toFilteredInputs node14Pkgs);

  shellHook = ''
    ${pkgs.lib.concatMapStrings (name: ''
      echo -n 12: ${name}:\ ; echo -en $(which ${name} 2>/dev/null)"\n" ;
    '') (pkgs.lib.attrNames node12Pkgs) }
    ${pkgs.lib.concatMapStrings (name: ''
      echo -n 14: ${name}:\ ; echo -en $(which ${name} 2>/dev/null)"\n" ;
    '') (pkgs.lib.attrNames node14Pkgs) }
  '';
}

```
</details>

\
:tada: 

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

\
cc @Mic92 @marsam @prusnak 
\
The broken `node-packages` need to be fixed, but i do not consider the fix being part of this PR.

### CODE REVIEW

File: pkgs/development/node-packages/node-packages.json [hunk 1]
```diff
@@ -3,7 +3,6 @@
 , "@antora/cli"
 , "@antora/site-generator-default"
 , "@bitwarden/cli"
-, "@vue/cli"
```
Comment (calbrecht) [inline pkgs/development/node-packages/node-packages.json:6]: i am adding those, right now. Don't know how they went missing.

File: pkgs/development/node-packages/node-packages.json [hunk 2]
```diff
@@ -35,7 +34,6 @@
 , "coc-python"
 , "coc-r-lsp"
 , "coc-rls"
-, "coc-rust-analyzer"
```
[the remark this answers is not in the document]
Comment (Mic92) [reply pkgs/development/node-packages/node-packages.json:38]: Not sure about this one, maybe it did not existed back than.

**Review by Mic92 — APPROVED** [review]
Looks good to me, if those missing packages gets re-added.

File: pkgs/development/node-packages/node-packages.json (hunk 1 above)
Comment (calbrecht) [reply pkgs/development/node-packages/node-packages.json]: Ok, done in https://github.com/NixOS/nixpkgs/pull/89184/commits/6149df8fd095cb6fe5a0a4fc2438e2805942ad59

File: pkgs/development/node-packages/node-packages.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (calbrecht) [reply pkgs/development/node-packages/node-packages.json]: Ok. Done in https://github.com/NixOS/nixpkgs/pull/89184/commits/6149df8fd095cb6fe5a0a4fc2438e2805942ad59

**Comment by calbrecht** [comment]
Opened #89205 to track/remove/fix broken packages.

**Comment by prusnak** [comment]
Love it! :+1:

**Comment by prusnak** [comment]
Since we are regenerating node-packages.nix, can we also change hashes from sha1 to sha256? This would fix most of the problems from https://github.com/NixOS/nixpkgs/issues/77238

**Comment by Mic92** [comment]
Those are coming from npm and are not computed locally. Hence we have little control over that.

**Comment by calbrecht** [comment]
@prusnak, yes @Mic92 is right, that would be something that has to be tackled in `node2nix` with, as there is already some abstraction over generating/converting [sha](https://github.com/svanderburg/node2nix/search?q=sha1&unscoped_q=sha1) sums.

Also i did the following to check the status:
```
tutnix ~/ws/nixpkgs-upstream (git) 6149df8fd095 U continue-pull-85764
% rg -o -c 'sha1' pkgs/development/node-packages/node-packages.nix
2342
tutnix ~/ws/nixpkgs-upstream (git) 6149df8fd095 U continue-pull-85764
% rg -o -c 'sha256' pkgs/development/node-packages/node-packages.nix
18
tutnix ~/ws/nixpkgs-upstream (git) 6149df8fd095 U continue-pull-85764
% rg -o -c 'sha512' pkgs/development/node-packages/node-packages.nix
3399
```

**Comment by Mic92** [comment]
Also the node ecosystem has bigger problem than sha1 collisions.

**Comment by calbrecht** [comment]
> Also the node ecosystem has bigger problem than sha1 collisions, i.e. malware.

Indeed, but i guess that holds true for any language specific package repository e.g. [python](https://www.zdnet.com/article/malicious-python-libraries-targeting-linux-servers-removed-from-pypi/)

**Comment by Mic92** [comment]
True also there we have more control/supervision since we handcraft all python modules.

**Comment by calbrecht** [comment]
Thank you @prusnak for taking care of that. I totally missed that.

**Pull request merged**